File: VB/MainDemo.Module/BusinessObjects/EmployeeTask.vb
Imports System
Imports DevExpress.Xpo
Imports System.ComponentModel
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Model
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Persistent.Base.General
Imports DevExpress.Persistent.Validation

Namespace MainDemo.Module.BusinessObjects
	<DefaultClassOptions, ImageName("BO_Task"), DefaultProperty("Subject"), ModelDefault("Caption", "Task")>
	Public Class EmployeeTask
		Inherits BaseObject

		Private _Description As String
		Private _Subject As String
		Private _DueDate As Date
		Private _Status As TaskStatus
		Private _AssignedTo As Employee
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		<Size(255)>
		Public Property Subject() As String
			Get
				Return _Subject
			End Get
			Set(ByVal value As String)
				SetPropertyValue("Subject", _Subject, value)
			End Set
		End Property
		Public Property Status() As TaskStatus
			Get
				Return _Status
			End Get
			Set(ByVal value As TaskStatus)
				SetPropertyValue("Status", _Status, value)
			End Set
		End Property
		<Size(SizeAttribute.Unlimited)>
		Public Property Description() As String
			Get
				Return _Description
			End Get
			Set(ByVal value As String)
				SetPropertyValue("Description", _Description, value)
			End Set
		End Property
		Public Property DueDate() As Date
			Get
				Return _DueDate
			End Get
			Set(ByVal value As Date)
				SetPropertyValue("DueDate", _DueDate, value)
			End Set
		End Property
		<Action(ToolTip := "Postpone the task to the next day", ImageName := "State_Task_Deferred")>
		Public Sub Postpone()
			If DueDate = Date.MinValue Then
				DueDate = Date.Now
			End If
			DueDate = DueDate.Add(TimeSpan.FromDays(1))
		End Sub
		<Association, RuleRequiredField>
		Public Property AssignedTo() As Employee
			Get
				Return _AssignedTo
			End Get
			Set(ByVal value As Employee)
				SetPropertyValue("AssignedTo", _AssignedTo, value)
			End Set
		End Property
	End Class
End Namespace

File: CS/MainDemo.Module/BusinessObjects/EmployeeTask.cs
using System;
using DevExpress.Xpo;
using System.ComponentModel;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp.Model;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Persistent.Base.General;
using DevExpress.Persistent.Validation;

namespace MainDemo.Module.BusinessObjects {
    [DefaultClassOptions]
    [ImageName("BO_Task")]
    [DefaultProperty("Subject")]
    [ModelDefault("Caption", "Task")]
    public class EmployeeTask : BaseObject {
        private string _Description;
        private string _Subject;
        private DateTime _DueDate;
        private TaskStatus _Status;
        private Employee _AssignedTo;
        public EmployeeTask(Session session)
            : base(session) {
        }
        [Size(255)]
        public string Subject {
            get {
                return _Subject;
            }
            set {
                SetPropertyValue("Subject", ref _Subject, value);
            }
        }
        public TaskStatus Status {
            get {
                return _Status;
            }
            set {
                SetPropertyValue("Status", ref _Status, value);
            }
        }
        [Size(SizeAttribute.Unlimited)]
        public string Description {
            get {
                return _Description;
            }
            set {
                SetPropertyValue("Description", ref _Description, value);
            }
        }
        public DateTime DueDate {
            get {
                return _DueDate;
            }
            set {
                SetPropertyValue("DueDate", ref _DueDate, value);
            }
        }
        [Action(ToolTip = "Postpone the task to the next day", ImageName = "State_Task_Deferred")]
        public void Postpone() {
            if(DueDate == DateTime.MinValue) {
                DueDate = DateTime.Now;
            }
            DueDate = DueDate + TimeSpan.FromDays(1);
        }
        [Association]
		[RuleRequiredField]
        public Employee AssignedTo {
            get {
                return _AssignedTo;
            }
            set {
                SetPropertyValue("AssignedTo", ref _AssignedTo, value);
            }
        }
    }
}
